packages/components/src/components/files/WebDav.ts

```diff
@@ -0,0 +1,151 @@
+        params ? super(params) : super();
+    }
+
+    static create(values): WebDavFile {
+        const webDavFile = new WebDavFile(values);
+
+        return webDavFile.merge({
+            isCollection: values.collection,
+            isLeaf: values.leaf,
+            createdBy: values.createdby,
+            created: values.creationdate,
+            lastModified: values.lastmodified,
+            downloadUrl: values.href ? values.href + '?contentDisposition=attachment' : undefined,
+            name: values.text,
+        }) as WebDavFile;
+    }
+}
+
+function getWebDavUrl(containerPath: string, directory?: string, createIntermediates?: boolean) {
+    let url =
+        ActionURL.getContextPath() +
+        '/_webdav' +
+        ActionURL.encodePath(containerPath) +
+        '/' +
+        encodeURIComponent('@files');
+
+    if (directory) {
+        url += '/' + directory;
+    }
+
+    if (createIntermediates) {
+        url += '?createIntermediates=' + createIntermediates;
+    }
+
+    return url;
+}
+
+export function getWebDavFiles(
+    containerPath: string,
+    directory?: string,
+    includeDirectories?: boolean
+): Promise<Map<string, WebDavFile>> {
+    return new Promise((resolve, reject) => {
+        const url = getWebDavUrl(containerPath, directory);
+
+        return Ajax.request({
+            url: url + '?method=JSON',
+            method: 'GET',
+            success: Utils.getCallbackWrapper(response => {
+                // Filter directories and create webdav files
+                const filteredFiles = response.files.reduce((filtered, file) => {
+                    if (includeDirectories || !file.collection) {
+                        return filtered.set(file.text, WebDavFile.create(file));
+                    }
+                }, Map<string, WebDavFile>());
+
+                resolve(filteredFiles);
+            }),
+            failure: Utils.getCallbackWrapper(
+                response => {
+                    console.error('Problem retrieving webDav files for container ' + containerPath);
+                    reject(response);
+                },
+                null,
+                false
+            ),
+        });
+    });
+}
+
+export function uploadWebDavFile(
+    file: File,
+    containerPath: string,
+    directory?: string,
+    createIntermediates?: boolean
+): Promise<string> {
+    return new Promise((resolve, reject) => {
+        const form = new FormData();
+        form.append('file', file);
+
+        const url = getWebDavUrl(containerPath, directory, createIntermediates);
+
+        Ajax.request({
+            url,
+            method: 'POST',
+            form,
+            success: Utils.getCallbackWrapper(() => {
+                resolve(file.name);
+            }),
+            failure: Utils.getCallbackWrapper(
+                () => {
+                    console.error('failure uploading file ' + file.name);
+                    reject(file.name);
```

Comment:
Are we sure we don't want to do anything with the server response to indicate what may have caused the failure?

Reply:
The only thing we get back here from the server is a blob of html.  Not really easy to get anything useful

Reply:
👍